### PR TITLE
ShadowMap fade option

### LINKED ISSUE #9564: Disable shadowMap fading
The shadow map is fading when the light source is low in the horizon.
Though this is a desirable behavior on earth, and bodies in general, where there is an atmosphere it does not make very much sense on other celestial bodies where no atmosphere is present like the Moon.
When the sun is low in the horizon the shadowMap darkness should be preserved.

I'm attaching a simple sandcastle example for demonstrating the current behavior: advance the time through the timeline for seeing the shadows fading

[sandcastle](https://sandcastle.cesium.com/#c=nVVRb9s2EP4rhF/iAjYlJU6bqW6wtgnSDMlazN4CDAYCWjpLXChSICm7aZH/3qMoyZbm7mF6kXj87uPdfcfTlmkCQvDSKJ6Sd0TCjnwEw6uCXrfmcfTm7M0sDCeD96u3K7mSQzR9uFlczJCqo0XYFo8pWPlFq38gsVzJ/lE3oDLNypwne8S48/fnOIpMqDUMXJ2pj61R1ORqd6NVJdP3tlCmzEE71w0TBt6SIECSBIxh+plslCaWCzDEKuL8CJdEKjkFpm2+T6SlBsnWAu54llsuMyS1uoIuSAtaMy4xky1PQffD9a9lHzL+vpIEn0qLmJzk1pYmDoKNi4iyxGbMAk1UETTMgagk04+ZSKMwPJl43y7I+DDel33tVGkNxtIc1RMj7i8bxjrX2L8a0yC1eGhoYObpeV/z2Je83cpZqnYmrkvWZ120W02p/PpepUCvf3//4e76auJycR6o3ic8j1RyL+Kft80JIHwatzLlCbNK9yNYMwN37Bn0F548wWDT8gIEl9CLLwOVqHQI5XKjPqivg/QSkOBCPkbOJC+Yr3drfunk2XLYDZvlr9o2Pknq5UclLZYJ9MmkFtNJ671ofSzdqKxpzbRrdE+O52agn+u0XRM0bj0zQntrytL01hu6Pj2I7QHW96xcgN7yBIa4fT+Tg4bewRr7jAqtEspMRSGtgrZ9hQ8N0a65H3cseXSNx0QLKJlmBVgPavhrxTSTBjdBWtJTzT14sbHijrVOLShlduJ3XzrBUe5skeRQoOhHR9LyADL+fvSikZdXez544KnNY3IenR4YP4EbF7XVXUt/L1vx/a3Aeu616UwI7L5pwb7yoiquuLFMJm6e4QjAh4b1SCsOsSnTTxIvB4L8rjO0F5DgCC6UamLAzQXY+miy4zjycmApJj0hJbdJjp2bEq2E6BouQS00owas61GvdwoGx2HT4O24wwGKX0ye0Y1WxRVkGsCMW4WmUUTPo06w6eyM/jKbdevo7Az/Nu3q+I/G7/ryK82xDZoImh5pMukiumc2p1b9gVZsnHE0o+etdnWuPwNOZ6f0okW6WvwMiFp4mGuxl1bntm5CJU8UxdN2iaMGpWlIfqsEel+5Se8KdWvUxeswGq9Gp+FpNA3Pp6ezZTSLwzB2Yofh36vRwf1vaVX5P1gv/ps1qbS7XA3x8Tz2Pu0Apd+UKpZqfBw/IUcDd2ePJqO5sc8CLlvdf+VFqbR102RMKf4Ei1JgRiZYVzhjLU1MPQoddB4cus5TviU8fbcaDQboakQSwYzBnU0lxIJ/g9Xoch4g/l+uQtX983kLGmeUg+XR5Z03UkrnAS6Pe1qlxJrpAfMP)

Current behavior
animation

Desired behavior
shadow_map_darkness_no_fading

Browser:
any

Operating System:
any

I already implemented this feature/fix and will push a pull request in a few

### PR DESCRIPTION
**Pull request by kalosma**
Fixes #9564

Added an option to shadowMap (`shadowMap.fadingEnabled`) for enabling/disabling the shadowMap fading when the light source is close to the horizon.
Having the fading disabled is desirable for cases, like the Moon, where there is no atmosphere and the shadows should be solid even if the light source is close to the horizon.

here is a sandcastle example with fading enabled (current behaviour, advance the time in timeline till the shadows fade)

[fading enabled (default)](https://sandcastle.cesium.com/#c=nVVRb9s2EP4rhF/iAjYlJU6bqW6wtgnSDMlazN4CDAYCWjpLXChSICm7aZH/3qMoyZbm7mF6kXj87uPdfcfTlmkCQvDSKJ6Sd0TCjnwEw6uCXrfmcfTm7M0sDCeD96u3K7mSQzR9uFlczJCqo0XYFo8pWPlFq38gsVzJ/lE3oDLNypwne8S48/fnOIpMqDUMXJ2pj61R1ORqd6NVJdP3tlCmzEE71w0TBt6SIECSBIxh+plslCaWCzDEKuL8CJdEKjkFpm2+T6SlBsnWAu54llsuMyS1uoIuSAtaMy4xky1PQffD9a9lHzL+vpIEn0qLmJzk1pYmDoKNi4iyxGbMAk1UETTMgagk04+ZSKMwPJl43y7I+DDel33tVGkNxtIc1RMj7i8bxjrX2L8a0yC1eGhoYObpeV/z2Je83cpZqnYmrkvWZ120W02p/PpepUCvf3//4e76auJycR6o3ic8j1RyL+Kft80JIHwatzLlCbNK9yNYMwN37Bn0F548wWDT8gIEl9CLLwOVqHQI5XKjPqivg/QSkOBCPkbOJC+Yr3drfunk2XLYDZvlr9o2Pknq5UclLZYJ9MmkFtNJ671ofSzdqKxpzbRrdE+O52agn+u0XRM0bj0zQntrytL01hu6Pj2I7QHW96xcgN7yBIa4fT+Tg4bewRr7jAqtEspMRSGtgrZ9hQ8N0a65H3cseXSNx0QLKJlmBVgPavhrxTSTBjdBWtJTzT14sbHijrVOLShlduJ3XzrBUe5skeRQoOhHR9LyADL+fvSikZdXez544KnNY3IenR4YP4EbF7XVXUt/L1vx/a3Aeu616UwI7L5pwb7yoiquuLFMJm6e4QjAh4b1SCsOsSnTTxIvB4L8rjO0F5DgCC6UamLAzQXY+miy4zjycmApJj0hJbdJjp2bEq2E6BouQS00owas61GvdwoGx2HT4O24wwGKX0ye0Y1WxRVkGsCMW4WmUUTPo06w6eyM/jKbdevo7Az/Nu3q+I/G7/ryK82xDZoImh5pMukiumc2p1b9gVZsnHE0o+etdnWuPwNOZ6f0okW6WvwMiFp4mGuxl1bntm5CJU8UxdN2iaMGpWlIfqsEel+5Se8KdWvUxeswGq9Gp+FpNA3Pp6ezZTSLwzB2Yofh36vRwf1vaVX5P1gv/ps1qbS7XA3x8Tz2Pu0Apd+UKpZqfBw/IUcDd2ePJqO5sc8CLlvdf+VFqbR102RMKf4Ei1JgRiZYVzhjLU1MPQoddB4cus5TviU8fbcaDQboakQSwYzBnU0lxIJ/g9Xoch4g/l+uQtX983kLGmeUg+XR5Z03UkrnAS6Pe1qlxJrpAfMP)

shadow_map_darkness

here is a sandcastle example with the new option to disable the shadows fading

[fading disabled (through new option)](https://sandcastle.cesium.com/#c=nVVbb9s2FP4rjF/iAjYlJc6auWmwtHYTD7lh9hZgMBDQ0rHEhSIFkrKbFvnvO9TNluruYXqRePid+3eONkwTEIJnRvGIfCQStuQzGJ6ndFqL+8H70/cj3x903u8+LOVSdtH06Xp+PkJTjVmEbdBNyrJHrf6B0HIl266uQcWaZQkPd4h+o1/6cSZioVbQUXWiNrZAUZOo7bVWuYyubKpMloB2qmsmDHwgnodGQjCG6VeyVppYLsAQq4jTI1wSqeQQmLbJLpHaNEi2EnDL48RyGaNRq3NogrSgNeMSM9nwCHQ73PK1aEP635eS4JNrMSbHibWZGXve2kVEWWhjZoGGKvUqy57IJdPPsYgC3z8elLpNkOP9eN92tVOZNRhL5arVjHH7WFksch2Xr0rUSW3cFVQw8/K6q/m4LHl9lbBIbc24KFnb6ry+qkpVnu9UBHR6f/XpdjoZuFycBnbvBv2RXO6a+Oes8gCiTGMmIx4yq3Q7ghUzcMteQT/y8AU6l5anILiEVnwxqFBFXSiXa/VJfe2kF4IEF/Ih40zylJX1rsVvTXs2HLZdsvxVyPrHYXH8rKTFMoE+HhTNdK0ttWjhlq5VXFEzaoheGke/MejXIm1HgkqtJUZo60xZFM1KQcPTvdieYHXHsjnoDQ+hi9vxmewRegsr5BkVWoWUmZxClHs1fUUZGqIduZ+3LHx2xGOiBmRMsxRsCarsFx3TTBq8BGlJq2vuwcHGijurRWpeJuPj8vataTi2O56HCaTY9IMrabEH6X8/OGjk7d3OHjzxyCZjchac7AlvwK2LQurGspzLuvnlVGA9d71pRAhsvmnKvvI0TyfcWCZDt89wBeBD/WKlpfvYiOkXicOBoPLWCeoBJLiCU6XQP14cHR0V7/vpE3l4XMwe7snigUxmcxw6sriZkvnN1eThaU6+XE1m99f7TtYswuJMf2DdsrQ8B1skRLYcF2kCBXpAMm7DBOchIloJ0dA4xA5rRg1Yx/ySRREYXLLV2NRLFNcyfjF5StdapROINYDp130fBgE9CxoaDEen9NfRqDkHp6f4D6tPh39f5W3ZVKU5kquKoGJelUkT0R2zCbXqD5QiHfvBiJ7VjChy/RlwODqh5zXS1eJnQOxwCXPEfavZU9dNqPCFIiW0XeACwzZURn7PBWpP3P/DFWpm1PkvftBf9k78k2Donw1PRotgNPb9saOQ7/+97O1tldqsyv6H1fP/thrm2o1sZfhwHjudei3Tb0qlC9U/jB+Qg4E7371B78LYVwGXdd9/42mmtHU7qk8p/lrTTGBGxlvluLktDU2xYB30wttXvYj4hvDo47LXWcvLHgkFMwZv1rkQc/4Nlr3LCw/xP6gKVfDnYQMaN5+DJcHlbSmklF54eDysaZUSK6Y7lv8F)

shadow_map_darkness_no_fading

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @kalosma!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/shadowmap_fade_opt/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Specs/Scene/ShadowMapSpec.js [hunk 1]
```diff
@@ -941,6 +941,48 @@ describe(
       });
     });
 
+    it("disable shadows fading", function () {
```
Comment (ebogo1) [inline Specs/Scene/ShadowMapSpec.js:944]: It looks like `effectiveDarkness` only gets used in the specs.. I think this test can be rewritten similarly to the `"changes darkness"` spec above. 

Something like:

```
...
scene.shadowMap = new ShadowMap( { ... } );

// Render with light looking straight down
var shadowedColor = renderAndReadPixels();

// Move the light close to the horizon
lightCamera.lookAt(center, new Cartesian3(1.0, 0.0, 0.01));

// Render with faded shadows
var horizonShadowedColor = renderAndReadPixels();
expect(horizonShadowedColor).not.toEqual(shadowedColor);

// Render with unfaded shadows
scene.shadowMap.fadingEnabled = false;
renderAndCall(function (rgba) {
  expect(rgba).not.toEqual(horizonShadowedColor);
});
```

Then the `effectiveDarkness` getter can be removed from `ShadowMap`.

**Review by ebogo1 — COMMENTED** [review]
Thanks @kalosma! I confirmed these changes work in Sandcastle. A few things -

File: Specs/Scene/ShadowMapSpec.js (hunk 1 above)
Comment (kalosma) [reply Specs/Scene/ShadowMapSpec.js:944]: thanks @ebogo1 for you suggestions. I made the changes you suggested: removed the `effectiveDarkness` getter and updated the unit test. Let me know if I need to change anything else

**Comment by kalosma** [comment]
Had some problems with the unit test, but it seems it passed now

File: Specs/Scene/ShadowMapSpec.js [hunk 2]
```diff
@@ -941,6 +941,42 @@ describe(
       });
     });
 
+    it("disable shadows fade out", function () {
```
Comment (ebogo1) [inline Specs/Scene/ShadowMapSpec.js:944]:
```suggestion
    it("disables shadow fading", function () {
```

**Review by ebogo1 — COMMENTED** [review]
Thanks, this PR looks just about ready. Just a minor change in wording for the test case and a [comment](https://github.com/CesiumGS/cesium/pull/9565/files#r638980113) I left earlier that you might've missed (I unresolved it). Should be good to go after those changes.

File: CHANGES.md [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 - Added `FeatureDetection.supportsBigInt64Array`, `FeatureDetection.supportsBigUint64Array` and `FeatureDetection.supportsBigInt`.
 
+- Added `ShadowMap.fadingEnabled` to control whether shadows fade out when the light source is close to the horizon. [#9565](https://github.com/CesiumGS/cesium/pull/9565)
```
Comment (ebogo1) [inline CHANGES.md:9]: @kalosma My mistake; I missed this the first time - to be consistent with other changes could you reword this as: 
```suggestion
- Added `options.fadingEnabled` parameter to `ShadowMap` to control whether shadows fade out when the light source is close to the horizon. [#9565](https://github.com/CesiumGS/cesium/pull/9565)
```

Otherwise this is good to go; I'll merge once that's done. Thanks!

**Review by ebogo1 — COMMENTED** [review]
Thanks @kalosma. I pushed a couple changes to CHANGES.md but this is good to :+1:
